--- a/target/linux/ramips/image/Makefile
+++ b/target/linux/ramips/image/Makefile
@@ -955,6 +955,11 @@ define Image/Build/Profile/Default
 	$(call Image/Build/Profile/ArcherC20i,$(1))
 	$(call Image/Build/Profile/MicroWRT,$(1))
 endef
+define Image/Build/Profile/GL-MT-ALL
+	$(call Image/Build/Profile/GL-MT300A,$(1))
+	$(call Image/Build/Profile/GL-MT300N,$(1))
+	$(call Image/Build/Profile/GL-MT750,$(1))
+endef
 endif
 
 #
--- a/target/linux/ramips/mt7620/profiles/gli.mk
+++ b/target/linux/ramips/mt7620/profiles/gli.mk
@@ -34,3 +34,12 @@ define Profile/GL-MT750/Description
     Support for gl-mt750 Router
 endef
 $(eval $(call Profile,GL-MT750))
+
+define Profile/GL-MT-ALL
+	NAME:=GL-MT-ALL
+	PACKAGES:=kmod-usb-core kmod-usb-dwc2 kmod-usb2 kmod-usb-ohci kmod-mt76 kmod-mt7610e
+endef
+define Profile/GL-MT-ALL/Description
+	All profiles including GL-MT300A, GL-MT300N, GL-MT750
+endef
+$(eval $(call Profile,GL-MT-ALL))
