--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -543,6 +543,16 @@ config ATH79_MACH_GL_AR150
    select ATH79_DEV_USB
    select ATH79_DEV_WMAC
 
+config ATH79_MACH_GL_MIFI
+	bool "GL MIFI support"
+	select SOC_AR933X
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_USB
+	select ATH79_DEV_WMAC
+
 config ATH79_MACH_GL_AR300
    bool "GL_AR300 support"
    select SOC_AR934X
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -79,6 +79,7 @@ obj-$(CONFIG_ATH79_MACH_EPG5000)	+= mach
 obj-$(CONFIG_ATH79_MACH_ESR1750)	+= mach-esr1750.o
 obj-$(CONFIG_ATH79_MACH_F9K1115V2)	+= mach-f9k1115v2.o
 obj-$(CONFIG_ATH79_MACH_GL_AR150)	+= mach-gl-ar150.o
+obj-$(CONFIG_ATH79_MACH_GL_MIFI)	+= mach-gl-mifi.o
 obj-$(CONFIG_ATH79_MACH_GL_AR300)	+= mach-gl-ar300.o
 obj-$(CONFIG_ATH79_MACH_GL_DOMINO)	+= mach-gl-domino.o
 obj-$(CONFIG_ATH79_MACH_GL_INET)	+= mach-gl-inet.o
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -68,6 +68,7 @@ enum ath79_mach_type {
 	ATH79_MACH_EPG5000,		/* EnGenius EPG5000 */
 	ATH79_MACH_F9K1115V2,		/* Belkin AC1750DB */
 	ATH79_MACH_GL_AR150,	/* GL-AR150 support */
+	ATH79_MACH_GL_MIFI,	    /* GL-MIFI support */
 	ATH79_MACH_GL_AR300,		/* GL-AR300 */
 	ATH79_MACH_GL_DOMINO,		/* Domino */
 	ATH79_MACH_GL_INET,		/* GL-CONNECT GL-INET */
