--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -563,6 +563,16 @@ config ATH79_MACH_GL_AR300
    select ATH79_DEV_USB
    select ATH79_DEV_WMAC
 
+config ATH79_MACH_GL_AR300M
+   bool "GL_AR300M support"
+   select SOC_QCA953X
+   select ATH79_DEV_ETH
+   select ATH79_DEV_GPIO_BUTTONS
+   select ATH79_DEV_LEDS_GPIO
+   select ATH79_DEV_M25P80
+   select ATH79_DEV_USB
+   select ATH79_DEV_WMAC
+
 config ATH79_MACH_GL_DOMINO
    bool "DOMINO support"
    select SOC_AR933X
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -81,6 +81,7 @@ obj-$(CONFIG_ATH79_MACH_F9K1115V2)	+= ma
 obj-$(CONFIG_ATH79_MACH_GL_AR150)	+= mach-gl-ar150.o
 obj-$(CONFIG_ATH79_MACH_GL_MIFI)	+= mach-gl-mifi.o
 obj-$(CONFIG_ATH79_MACH_GL_AR300)	+= mach-gl-ar300.o
+obj-$(CONFIG_ATH79_MACH_GL_AR300M)    += mach-gl-ar300m.o
 obj-$(CONFIG_ATH79_MACH_GL_DOMINO)	+= mach-gl-domino.o
 obj-$(CONFIG_ATH79_MACH_GL_INET)	+= mach-gl-inet.o
 obj-$(CONFIG_ATH79_MACH_GS_MINIBOX_V1)	+= mach-gs-minibox-v1.o
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -70,6 +70,7 @@ enum ath79_mach_type {
 	ATH79_MACH_GL_AR150,	/* GL-AR150 support */
 	ATH79_MACH_GL_MIFI,	    /* GL-MIFI support */
 	ATH79_MACH_GL_AR300,		/* GL-AR300 */
+	ATH79_MACH_GL_AR300M,		/* GL-AR300M */
 	ATH79_MACH_GL_DOMINO,		/* Domino */
 	ATH79_MACH_GL_INET,		/* GL-CONNECT GL-INET */
 	ATH79_MACH_GS_MINIBOX_V1,	/* Gainstrong MiniBox V1.0 */
